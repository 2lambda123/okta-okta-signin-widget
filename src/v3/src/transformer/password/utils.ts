/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { PasswordWithConfirmationElement, UISchemaElement } from '../../types';

export const getPasswordMatchingKey = (
  data: Record<string, unknown>,
): string | undefined => {
  if ('credentials.passcode' in data) {
    return 'credentials.passcode';
  }

  if ('credentials.password' in data) {
    return 'credentials.password';
  }

  if ('credentials.newPassword' in data) {
    return 'credentials.newPassword';
  }

  // Should never hit this case as it should be one of values defined above
  return undefined;
};

export const updatePasswordWithConfirmationBtnLabel = (
  elements: UISchemaElement[],
  label: string,
): void => {
  const pwEnrollmentEle = elements.find(
    (element) => element.type === 'PasswordWithConfirmation',
  ) as PasswordWithConfirmationElement;
  if (pwEnrollmentEle) {
    pwEnrollmentEle.options = {
      ...pwEnrollmentEle.options,
      ctaLabel: label,
    };
  }
};
