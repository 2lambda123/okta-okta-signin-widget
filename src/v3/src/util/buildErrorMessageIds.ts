/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

export const buildErrorMessageIds = (errors: string[], fieldName: string): string => {
  if (errors.length === 1) {
    return `${fieldName}-error`;
  }
  const ids = errors.map((_: string, index: number) => `${fieldName}-error-${index}`);
  return ids.join(' ');
};
