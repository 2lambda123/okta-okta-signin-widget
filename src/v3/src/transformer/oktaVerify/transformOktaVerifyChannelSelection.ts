/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { loc } from 'okta';
import { IDX_STEP } from '../../constants';
import {
  ButtonElement,
  ButtonType,
  FieldElement,
  IdxStepTransformer,
  TitleElement,
  UISchemaElement,
} from '../../types';
import { isAndroidOrIOS } from '../../util';
import { getUIElementWithName } from '../utils';

export const CHANNEL_SCOPE = '#/properties/authenticator/properties/channel';
const CHANNEL_TO_LABEL_KEY_MAP: { [channel: string]: string } = {
  qrcode: 'oie.enroll.okta_verify.select.channel.qrcode.label',
  email: 'oie.enroll.okta_verify.select.channel.email.label',
  sms: 'oie.enroll.okta_verify.select.channel.sms.label',
};

export const transformOktaVerifyChannelSelection: IdxStepTransformer = ({
  prevTransaction,
  formBag,
}) => {
  const { data, uischema } = formBag;

  const isAndroidOrIOSView = isAndroidOrIOS();
  // eslint-disable-next-line max-len
  const lastSelectedChannel = prevTransaction?.context.currentAuthenticator.value.contextualData?.selectedChannel as string;
  const elements: UISchemaElement[] = [];

  const titleElement: TitleElement = {
    type: 'Title',
    options: {
      content: isAndroidOrIOSView
        ? loc('oie.enroll.okta_verify.setup.title', 'login')
        : loc('oie.enroll.okta_verify.select.channel.title', 'login'),
    },
  };
  elements.push(titleElement);

  const channelSelectionElement = getUIElementWithName(
    'authenticator.channel',
    formBag.uischema.elements as UISchemaElement[],
  ) as FieldElement;
  channelSelectionElement.label = loc('oie.enroll.okta_verify.select.channel.description', 'login');
  channelSelectionElement.options.format = 'radio';
  const { options: { inputMeta: { options = [] } } } = channelSelectionElement;
  channelSelectionElement.options.customOptions = options
    .filter((opt) => opt.value !== lastSelectedChannel)
    .map((opt) => ({
      value: opt.value as string,
      label: loc(CHANNEL_TO_LABEL_KEY_MAP[opt.value as string], 'login'),
    }));
  elements.push(channelSelectionElement);

  data['authenticator.channel'] = channelSelectionElement.options.customOptions[0].value;

  const submitButton: ButtonElement = {
    type: 'Button',
    label: loc('oform.next', 'login'),
    scope: `#/properties/${ButtonType.SUBMIT}`,
    options: {
      type: ButtonType.SUBMIT,
      step: IDX_STEP.SELECT_ENROLLMENT_CHANNEL,
    },
  };
  elements.push(submitButton);

  if (!['email', 'sms'].includes(lastSelectedChannel)) {
    const switchChannelButton = {
      type: 'Button',
      label: loc('next.enroll.okta_verify.switch.channel.link.text', 'login'),
      options: {
        type: ButtonType.BUTTON,
        step: 'select-enrollment-channel',
        variant: 'secondary',
      },
    };
    elements.push(switchChannelButton);
  }

  uischema.elements = elements;

  return formBag;
};
