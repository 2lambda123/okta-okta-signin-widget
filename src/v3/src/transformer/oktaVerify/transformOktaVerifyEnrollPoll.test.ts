/*
 * Copyright (c) 2022-present, Okta, Inc. and/or its affiliates. All rights reserved.
 * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
 *
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *
 * See the License for the specific language governing permissions and limitations under the License.
 */

import { IdxContext } from '@okta/okta-auth-js';
import { getStubTransactionWithNextStep } from 'src/mocks/utils/utils';
import {
  DescriptionElement,
  FormBag,
  ListElement,
  ReminderElement,
  StepperButtonElement,
  StepperLayout,
  TitleElement,
  UISchemaLayout,
  UISchemaLayoutType,
  WidgetProps,
} from 'src/types';

import * as utils from '../../util/browserUtils';
import * as channelTransformer from './transformOktaVerifyChannelSelection';
import { transformOktaVerifyEnrollPoll } from './transformOktaVerifyEnrollPoll';

describe('TransformOktaVerifyEnrollPoll Tests', () => {
  const transaction = getStubTransactionWithNextStep();
  let formBag: FormBag;
  let channelSelectFormBag: FormBag;
  let mobileDeviceStub: jest.SpyInstance<boolean>;
  let channelTransformerStub: jest.SpyInstance;
  let mockProps: WidgetProps;

  beforeEach(() => {
    formBag = {
      schema: {},
      uischema: {
        type: UISchemaLayoutType.VERTICAL,
        elements: [],
      },
      data: {},
    };

    channelSelectFormBag = {
      schema: {},
      uischema: {
        type: UISchemaLayoutType.VERTICAL,
        elements: [{ type: 'Title', options: { content: 'Select a channel' } } as TitleElement],
      },
      data: {},
    };

    mobileDeviceStub = jest.spyOn(utils, 'isAndroidOrIOS');
    channelTransformerStub = jest.spyOn(
      channelTransformer,
      'transformOktaVerifyChannelSelection',
    ).mockReturnValue(channelSelectFormBag);
    mockProps = {};
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should return a channel selection formBag when mobile device and qrcode is selected channel', () => {
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: { selectedChannel: 'qrcode' },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(true);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    expect(updatedFormBag).toStrictEqual(channelSelectFormBag);

    expect(mobileDeviceStub).toBeCalledTimes(1);
    expect(channelTransformerStub).toBeCalledTimes(1);
    mobileDeviceStub.mockRestore();
    channelTransformerStub.mockRestore();
  });

  it('should return a channel selection formBag when on mobile device and sms is selected channel with canResend = true', () => {
    transaction.availableSteps = [{ name: 'resend' }];
    transaction.nextStep = {
      name: '',
      canResend: true,
    };
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: {
            selectedChannel: 'sms',
            phoneNumber: '+14215551262',
          },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(true);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(4);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout)
        .elements[0] as UISchemaLayout)
        .elements[0] as ReminderElement).options?.ctaText,
    ).toBe('next.enroll.okta_verify.sms.notReceived');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.setup.sms.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('next.enroll.okta_verify.sms.info');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.contentParams,
    ).toEqual(['+14215551262']);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).label,
    ).toBe('next.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });

  it('should return a channel selection formBag when on mobile device and email is selected channel and canResend = true', () => {
    transaction.availableSteps = [{ name: 'resend' }];
    transaction.nextStep = {
      name: '',
      canResend: true,
    };
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: {
            selectedChannel: 'email',
            email: 'noreply@noemail.com',
          },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(true);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(4);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as ReminderElement).options?.ctaText,
    ).toBe('next.enroll.okta_verify.email.notReceived');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.setup.email.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('next.enroll.okta_verify.email.info');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.contentParams,
    ).toEqual(['noreply@noemail.com']);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).label,
    ).toBe('next.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });

  it('should add Stepper elements when selectedChannel is qrcode and not on mobile device and canResend = true', () => {
    transaction.availableSteps = [{ name: 'resend' }];
    transaction.nextStep = {
      name: '',
      canResend: true,
    };
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: {
            selectedChannel: 'qrcode',
            qrcode: { href: '', method: '', type: '' },
          },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(false);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(4);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.setup.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as ListElement).options?.items,
    ).toEqual([
      'oie.enroll.okta_verify.qrcode.step1',
      'oie.enroll.okta_verify.qrcode.step2',
      'oie.enroll.okta_verify.qrcode.step3',
    ]);
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2].type,
    ).toBe('QRCode');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).label,
    ).toBe('enroll.totp.cannotScan');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(2);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[1] as StepperButtonElement).label,
    ).toBe('next.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[1] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[1] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(0);
  });

  it('should add Stepper elements when selectedChannel is sms and not on mobile device and canResend = true', () => {
    transaction.availableSteps = [{ name: 'resend' }];
    transaction.nextStep = {
      name: '',
      canResend: true,
    };
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: {
            selectedChannel: 'sms',
            phoneNumber: '+14215551262',
          },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(false);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(4);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as ReminderElement).options?.ctaText,
    ).toBe('next.enroll.okta_verify.sms.notReceived');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.setup.sms.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('next.enroll.okta_verify.sms.info');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.contentParams,
    ).toEqual(['+14215551262']);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).label,
    ).toBe('next.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });

  it('should add Stepper elements when selectedChannel is email and not on mobile device and canResend = true', () => {
    transaction.availableSteps = [{ name: 'resend' }];
    transaction.nextStep = {
      name: '',
      canResend: true,
    };
    transaction.context = {
      // TODO: OKTA-503490 temporary sln access missing relatesTo obj
      currentAuthenticator: {
        value: {
          contextualData: {
            selectedChannel: 'email',
            email: 'noreply@noemail.com',
          },
        },
      },
    } as unknown as IdxContext;
    mobileDeviceStub.mockReturnValue(false);

    const updatedFormBag = transformOktaVerifyEnrollPoll(transaction, formBag, mockProps);

    // Main Stepper
    expect(updatedFormBag.uischema.elements.length).toBe(1);
    // Steps
    expect((updatedFormBag.uischema.elements[0] as StepperLayout).elements.length).toBe(2);
    // First step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements.length,
    ).toBe(4);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[0] as ReminderElement).options?.ctaText,
    ).toBe('next.enroll.okta_verify.email.notReceived');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[1] as TitleElement).options?.content,
    ).toBe('oie.enroll.okta_verify.setup.email.title');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.content,
    ).toBe('next.enroll.okta_verify.email.info');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[2] as DescriptionElement).options?.contentParams,
    ).toEqual(['noreply@noemail.com']);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).label,
    ).toBe('next.enroll.okta_verify.switch.channel.link.text');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.variant,
    ).toBe('secondary');
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[0] as UISchemaLayout)
        .elements[3] as StepperButtonElement).options?.nextStepIndex,
    ).toBe(1);

    // Second step steps
    expect(
      ((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements.length,
    ).toBe(1);
    expect(
      (((updatedFormBag.uischema.elements[0] as StepperLayout).elements[1] as UISchemaLayout)
        .elements[0] as TitleElement).options?.content,
    ).toBe('Select a channel');
  });
});
