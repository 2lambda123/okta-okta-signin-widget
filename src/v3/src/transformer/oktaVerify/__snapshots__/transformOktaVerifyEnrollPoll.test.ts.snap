// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TransformOktaVerifyEnrollPoll Tests should add Stepper elements when selectedChannel is qrcode and canResend = true 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "items": Array [
                    "oie.enroll.okta_verify.qrcode.step1",
                    "oie.enroll.okta_verify.qrcode.step2",
                    "oie.enroll.okta_verify.qrcode.step3",
                  ],
                  "type": "ordered",
                },
                "type": "List",
              },
              Object {
                "options": Object {
                  "data": "",
                  "label": undefined,
                },
                "type": "QRCode",
              },
              Object {
                "label": "enroll.totp.cannotScan",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.email.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.sms.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "defaultStepIndex": [Function],
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`TransformOktaVerifyEnrollPoll Tests should return a channel selection formBag when email is selected channel and canResend = true 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.email.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.email.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "items": Array [
                    "oie.enroll.okta_verify.qrcode.step1",
                    "oie.enroll.okta_verify.qrcode.step2",
                    "oie.enroll.okta_verify.qrcode.step3",
                  ],
                  "type": "ordered",
                },
                "type": "List",
              },
              Object {
                "options": Object {
                  "data": undefined,
                  "label": undefined,
                },
                "type": "QRCode",
              },
              Object {
                "label": "enroll.totp.cannotScan",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.email.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.email.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.email.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.email.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.email.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.sms.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "defaultStepIndex": [Function],
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;

exports[`TransformOktaVerifyEnrollPoll Tests should return a channel selection formBag when sms is selected channel with canResend = true 1`] = `
Object {
  "data": Object {},
  "dataSchema": Object {},
  "schema": Object {},
  "uischema": Object {
    "elements": Array [
      Object {
        "elements": Array [
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.sms.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.sms.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "items": Array [
                    "oie.enroll.okta_verify.qrcode.step1",
                    "oie.enroll.okta_verify.qrcode.step2",
                    "oie.enroll.okta_verify.qrcode.step3",
                  ],
                  "type": "ordered",
                },
                "type": "List",
              },
              Object {
                "options": Object {
                  "data": undefined,
                  "label": undefined,
                },
                "type": "QRCode",
              },
              Object {
                "label": "enroll.totp.cannotScan",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.sms.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.sms.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.email.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
          Object {
            "elements": Array [
              Object {
                "options": Object {
                  "action": [Function],
                  "ctaText": "oie.enroll.okta_verify.sms.notReceived",
                  "linkLabel": "email.button.resend",
                },
                "type": "Reminder",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.setup.sms.title",
                },
                "type": "Title",
              },
              Object {
                "options": Object {
                  "content": "oie.enroll.okta_verify.sms.info",
                },
                "type": "Description",
              },
              Object {
                "label": "oie.enroll.okta_verify.switch.channel.link.text",
                "options": Object {
                  "step": "select-enrollment-channel",
                  "stepToRender": "select-enrollment-channel",
                  "type": "button",
                  "variant": "secondary",
                },
                "type": "Button",
              },
            ],
            "type": "VerticalLayout",
          },
        ],
        "options": Object {
          "defaultStepIndex": [Function],
        },
        "type": "Stepper",
      },
    ],
    "type": "VerticalLayout",
  },
}
`;
